Kuadrant/authorino-operator
# Authorino K8s auth ClusterRole

Add support for `authorino-manager-k8s-auth-role` `Role`.

Related to https://github.com/Kuadrant/authorino/pull/211

Plus:
- Fixes owner references of Authorino `Service`, `ServiceAccount` and other resources.
- Fixes `ClusterRoleBinding`s created to append new subjects instead of always creating another binding – prevents overriding the cluster-scoped resource when there are more than one instance of Authorino with the same name (in different k8s namespaces of course)
- ~Prepends the name of the `Authorino` CR's namespace to the name of the `ClusterRoleBinding`s created.~

TODO:
- [ ] ~Fix unique names of the `ClusterRoleBinding`s – `<cr-namespace>-<cr-name>-<crb-name>` is not enough to prevent clashes (e.g. `namespace: foo-x / name: bar` and `namespace: foo / name: x-bar`)~
- [x] Fix cleanup on deletion of `Authorino` custom resources – Due to the fix for appending subjects to the `ClusterRoleBinding` instead of creating new ones, the cleanup is now broken. When the `Authorino` resource is deleted, the corresponding `ServiceAccount` is not removed from the list of subjects of the `ClusterRoleBinding`s. To fix that, we can either
    - ~get back to creating one set of `ClusterRoleBinding`s per Authorino instance (adding the name and the namespace of the `Authorino` resource to the name of the `ClusterRoleBinding`, to avoid overwriting)~ – it will not work due to k8s ownerreferences not allowed between cluster-scoped dependents and namespaced owners ([ref](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/)), or
    - implement the removal of the subject from the `ClusterRoleBinding`s on 404 Not Found of `Authorino` CRs caught by the reconciler. [DONE]

### Verification steps

Build and deploy images of Authorino and Authorino Operator that contains the required changes:

```sh
cd $GOPATH/src/github.com/kuadrant/authorino-operator
git checkout authorino-k8s-auth-role
make docker-build OPERATOR_IMAGE=authorino-operator:local

cd $GOPATH/src/github.com/kuadrant/authorino
git checkout fix/k8s-auth-namespaced
make cluster
make cert-manager
make local-build # kind load docker-image authorino:local --name authorino
kind load docker-image authorino-operator:local --name authorino
make install-operator OPERATOR_VERSION=authorino-k8s-auth-role
kubectl -n authorino-operator patch deployment/authorino-operator-controller-manager -p '{"spec":{"template":{"spec":{"containers":[{"name":"manager","image":"authorino-operator:local"}]}}}}'
make install
make namespace
make deploy

kubectl -n authorino apply -f -<<EOF
apiVersion: operator.authorino.kuadrant.io/v1beta1
kind: Authorino
metadata:
  name: authorino-2
spec:
  clusterWide: false
  listener:
    tls:
      enabled: false
  oidcServer:
    tls:
      enabled: false
EOF
```

Check the resources created:

```sh
kubectl get clusterroles -o name | grep authorino-manager
# clusterrole.rbac.authorization.k8s.io/authorino-manager-k8s-auth-role
# clusterrole.rbac.authorization.k8s.io/authorino-manager-role

kubectl get clusterrolebindings | grep authorino-manager
# authorino-k8s-auth                                     ClusterRole/authorino-manager-k8s-auth-role                                        46s

kubectl get clusterrolebindings/authorino-k8s-auth -o yaml
# apiVersion: rbac.authorization.k8s.io/v1
# kind: ClusterRoleBinding
# metadata:
#   creationTimestamp: "2022-01-12T12:08:58Z"
#   name: authorino-k8s-auth
#   resourceVersion: "3734"
#   uid: 3c99dc92-2173-4c4c-bec5-0d94906dd8d7
# roleRef:
#   apiGroup: rbac.authorization.k8s.io
#   kind: ClusterRole
#   name: authorino-manager-k8s-auth-role
# subjects:
# - kind: ServiceAccount
#   name: authorino-authorino
#   namespace: authorino
# - kind: ServiceAccount
#   name: authorino-2-authorino
#   namespace: authorino

kubectl -n authorino get serviceaccounts -o name | grep authorino
# serviceaccount/authorino-2-authorino
# serviceaccount/authorino-authorino

kubectl -n authorino get rolebindings | grep authorino
# authorino-2-authorino                   ClusterRole/authorino-manager-role    106s
# authorino-2-authorino-leader-election   Role/authorino-leader-election-role   106s
# authorino-authorino                     ClusterRole/authorino-manager-role    3m49s
# authorino-authorino-leader-election     Role/authorino-leader-election-role   3m49s
```

Delete one of the Authorino instances and check again:

```sh
kubectl -n authorino delete authorinos/authorino
# authorino.operator.authorino.kuadrant.io "authorino" deleted

kubectl get clusterrolebindings | grep authorino-manager
# authorino-k8s-auth                                     ClusterRole/authorino-manager-k8s-auth-role                                        4m58s

kubectl get clusterrolebindings/authorino-k8s-auth -o yaml
# apiVersion: rbac.authorization.k8s.io/v1
# kind: ClusterRoleBinding
# metadata:
#   creationTimestamp: "2022-01-12T12:08:58Z"
#   name: authorino-k8s-auth
#   resourceVersion: "3734"
#   uid: 3c99dc92-2173-4c4c-bec5-0d94906dd8d7
# roleRef:
#   apiGroup: rbac.authorization.k8s.io
#   kind: ClusterRole
#   name: authorino-manager-k8s-auth-role
# subjects:
# - kind: ServiceAccount
#   name: authorino-2-authorino
#   namespace: authorino
```
